### PR TITLE
FEATURE: Adding support for users to deploy Featureform on docker via cli. 

### PR DESCRIPTION
**Pull request by ahmadnazeri**
# Description
This PR will add support for deploying both Featureform as well as quickstart via `featureform` cli. 

Deploying Featureform on Docker: 
```shell
featureform deploy docker
```

Deploying Featureform and Quickstart on Docker:
```shell
featureform deploy docker --quickstart
```

Output of when containers are already running: 

```shell
Deploying Featureform on Docker
Starting Docker deployment on Darwin 22.5.0
Checking if featureform container exists...
        Container featureform is already running. Skipping...
Checking if quickstart-postgres container exists...
        Container quickstart-postgres is already running. Skipping...
Checking if quickstart-redis container exists...
        Container quickstart-redis is already running. Skipping...

Pulling Quickstart files
        Pulling definitions.py
                definitions.py already exists. Skipping...

Featureform is now running!
export FEATUREFORM_HOST=http://localhost:7878
To access the dashboard, visit http://localhost:80
```

Also, we now support teardown as well. 

```shell
featureform teardown docker
```

Output:
```shell
Tearing down Featureform on Docker
Stopping containers...
        Stopping featureform container
        Stopping quickstart-postgres container
        Stopping quickstart-redis container
```

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# Checklist:

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: client/src/featureform/cli.py [hunk 1]
```diff
@@ -249,6 +254,25 @@ def search(query, host, cert, insecure, local):
...
+    default="docker",
+    type=click.Choice(["docker", "kubernetes"]),
+)
+@click.option("--quickstart", is_flag=True, help="Install Featureform Quickstart as well")
+def deploy(deploy_type, quickstart):
+    print("Deploying Featureform...", quickstart, deploy_type)
+    if deploy_type == "docker":
+        deployment = DockerDeployment(quickstart)
+    else:
+        raise ValueError("Invalid deployment type: Supported types are 'docker'")
```
Comment (anthonylasso) [inline client/src/featureform/cli.py:270]: is kubernetes just a poc? it's a listed option on 262.

Comment (anthonylasso) [reply client/src/featureform/cli.py:270]: sidebar chat

File: Dockerfile [hunk 2]
```diff
@@ -128,4 +128,6 @@ ENV MEILI_LOG_LEVEL="WARN"
 EXPOSE 7878
 EXPOSE 80
 
+HEALTHCHECK --interval=5m --timeout=10s --start-period=10s --retries=3 CMD curl --fail http://localhost/ || exit 1
```
Comment (aolfat) [inline Dockerfile:131]: what this do?

File: client/src/featureform/deploy.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+
+warnings.simplefilter("ignore")
+
+DOCKER_CONFIG = namedtuple(
+    "Docker_Config", ["name", "image", "port", "detach_mode", "env"]
+)
+
+
+class Deployment:
```
Comment (aolfat) [inline client/src/featureform/deploy.py:17]: nit: I would make this an abstract class, we don't ever want to instatiate one of these directly

Comment (aolfat) [inline client/src/featureform/deploy.py:17]: you can just make this a dataclass

File: client/src/featureform/deploy.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    def __init__(self, quickstart: bool):
+        self._quickstart = quickstart
+        self._status = None
+        self._config = []
```
Comment (aolfat) [inline client/src/featureform/deploy.py:21]: nit: self._docker_configs = []

File: client/src/featureform/deploy.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+        featureform_deployment = DOCKER_CONFIG(
+            name="featureform",
+            image=os.getenv(
+                "FEATUREFORM_DOCKER_IMAGE", "featureformcom/featureform:latest"
+            ),
+            port={"7878/tcp": 7878, "80/tcp": 80},
+            detach_mode=True,
+            env=environment_variables,
+        )
+
```
Comment (aolfat) [inline client/src/featureform/deploy.py:71]: self._docker_configs.append(featureform_deployment)

File: client/src/featureform/deploy.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+                name="quickstart-redis",
+                image="redis:latest",
+                port={"6379/tcp": 6379},
+                detach_mode=True,
+                env={},
+            ),
+        ]
+
+        if self._quickstart:
+            self._config = [featureform_deployment] + quickstart_deployment
```
Comment (aolfat) [inline client/src/featureform/deploy.py:91]: self._docker_configs.append(quickstart_deployment)

**Comment by aolfat** [comment]
I'm ok with this, I don't think we need to sit down and make sure it's perfect as it's just a script, as long as it works I'm cool with it

**Review by aolfat — APPROVED** [review]

File: Dockerfile (hunk 2 above)
Comment (ahmadnazeri) [reply Dockerfile:131]: if 5 minutes, it hits localhost to see if it is up. If it isn't then the container will fail.

File: client/src/featureform/deploy.py (hunk 4 above)
Comment (ahmadnazeri) [reply client/src/featureform/deploy.py:21]: well, it doesn't have to be docker.

**Pull request merged**